src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,370 @@
+            {
+                _readHead = null;
+                _commitHead = null;
+                returnEnd = null;
+                _readIndex = 0;
+            }
+            else if (consumedIndex == returnEnd.Length)
+            {
+                var nextBlock = returnEnd.NextSegment;
+                _readHead = nextBlock;
+                _readIndex = 0;
+                returnEnd = nextBlock;
+            }
+            else
+            {
+                _readHead = consumedSegment;
+                _readIndex = consumedIndex;
+            }
+
+            // Remove all blocks that 
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnStart.ResetMemory();
+                returnStart = returnStart.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            if (_isCompleted)
+            {
+                throw new InvalidOperationException("Reading is not allowed after reader was completed.");
+            }
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
+            {
+                return readResult;
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+            using (reg)
+            {
+                try
+                {
+                    AllocateCommitHead();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_commitHead.AvailableMemory, tokenSource.Token);
+#elif NETSTANDARD2_0
+                    MemoryMarshal.TryGetArray<byte>(_commitHead.AvailableMemory, out var arraySegment);
```

Comment:
What if this returns `false`?

Reply:
NETSTANDARD2_0 is getting removed shortly anyways.

Reply:
We need to keep netstandard here if we want to port this to pipelines.